app/services/notification_service.rb

```diff
@@ -56,6 +56,17 @@ def notify(object, recipient_user_ids)
     notification_types(object).each {|type| type.notify(object, recipient_user_ids) }
   end
 
+  def read_all_involving(person)
+    Notification
+      .for(@user)
+      .joins(:notification_actors)
+      .where(notification_actors: {person: person})
+      .find_each do |note|
+        note.unread = false
+        note.save
+      end
```

Comment:
I'm wondering if it would be worth also filtering only still unread notifications? Or instead of changing each entity separately, maybe use `update_all(unread: false)`? This would skip the vailidation of the entity (as it directly calls the database and skips the rails logic) and pronto/rubocop will be unhappy about it. But there is no validation based on the read-state, if it was valid before, it's also still valid after being marked as read, so I think in this case it should be fine to ignore this rule.

However, I think the current implementation is fine for now for the current intended use-case, where people probably don't have a lot of already read notifications for a person they want to block. And also they probably don't have a lot of notifications overall yet for the person they want to block (at least not as many as it would make a notable performance difference).

Reply:
Ahh, insightful, thank you! I went with filtering the already read notes in the `where`.

I don't like sending a query that updates a row with exactly the same data, I didn't realize that Rails' optimizations didn't reach that far. Also, thinking about the use case, I'm not sure, but think that it will probably happen sometimes that a user wants to block another user with whom they have have been interacting for a while.

About the validations, I like the idea of not skipping them right now, because I want to make it easy to extend them in the future (therefore reducing the amount of places where they can be skipped).

Reply:
> after more research, maybe Rails doesn't send a query if the record didn't change upon save, but it will run the validations.

I'm also not 100% sure if rails really sends out the update or what it maybe automatically skips, but it will at least run something, and for a loop with potentially a lot of entries, I at least wanted to think about.

> About the validations, I like the idea of not skipping them right now, because I want to make it easy to extend them in the future (therefore reducing the amount of places where they can be skipped).

Yeah, I also thought about if it could be a problem for future validations, but I couldn't think of any potential useful validation where a notification could be valid if it's unread, but then invalid if it's read. But I'm fine with leaving it with the loop and update each notification separately. The `update_all` was only an idea that would enable it to do it with just one query. But with the read notifications filtered out, there shouldn't be a lot of notifications left in most cases, so this is fine. :+1: